content/blog/2020-05-updates-to-ragg-and-systemfonts/index.Rmarkdown

```diff
@@ -0,0 +1,180 @@
+  text, 
+  id = 1, 
+  family = family,
+  italic = italic,
+  bold = bold,
+  size = size,
+  tracking = c(0, 400, rep(0, 4)),
+  width = 4.5,
+  indent = 0.2,
+  hjust = 0.5,
+  vjust = 0.5,
+  res = 300
+)
+
+shape
+```
+
+```{r}
+library(grid)
+
+pngfile <- path(knitr::fig_path(),  "png_image.png")
+agg_png(pngfile, 6, 2.5, unit = 'in', res = 300)
+grid.text(
+  shape$shape$glyph, 
+  x = unit(0.5, "npc") + unit(shape$shape$x_offset, "pt"), 
+  y = unit(0.5, "npc") + unit(shape$shape$y_offset, "pt"),
+  hjust = 0,
+  vjust = 0,
+  gp = gpar(
+    fontfamily = family[shape$shape$string_id],
+    fontface = style[shape$shape$string_id],
+    fontsize = size[shape$shape$string_id]
+  )
+)
+invisible(dev.off())
+knitr::include_graphics(pngfile)
+```
+
+When you see this you may immediately think of the work Claus Wilke has done with [gridtext](https://wilkelab.org/gridtext/). We are both committed to improving the rich text support in R, and are attacking the problem from different directions. Claus has worked on doing per-word shaping, whereas the text shaping done in systemfonts are done glyph-by-glyph. Both approaches has their merits, and we are currently at a stage where we are trying to figure out how to best solve this in R. At the time of writing there is no doubt that gridtext is more complete and stable, but doing glyph-based shaping offers possibilities that are not possible with gridtext, e.g. setting tracking and flowing text along paths. All of this is highly experimental still, and there will need to be some changes in the graphic engine to make glyph-based rendering performant.
+
+### The font-ier
+A big (for some) issue that I haven't touched upon is support for ligatures. A **ligature** is a glyph that represents multiple characters, as when "f" and "i" are substituted for the single glyph "`r "\UFB01"`". Ligatures are often used as flourishes in Western scripts but are essential for other scripts such as Arabic. Further, ligatures power emojis which have become an integral part of modern communication. 
+
+While we don't currently have support for ligatures, there's no doubt that we need to in order to be more inclusive in the script types that we support. The exact mechanism for this is still up in the air because ligatures must be resolved at the font rendering level. Each font file may provide their own unique ligatures and the mapping from multiple characters to a single glyph can be encoded into a number of different substitution table formats in the font file. Because of this complexity it is often handled by a dedicated library such as [HarfBuzz](https://www.freedesktop.org/wiki/Software/HarfBuzz/). This could of course be wrapped by an R package but since ligatures can not be encoded with unicode (for the most part), there is no way of precomputing the substitution and pass it on to the graphic devices. Whether ligature support can retroactively be added to old graphic devices is thus still up in the air.
```

Comment:
```suggestion
While we don't currently have support for ligatures, there's no doubt that we need. The exact mechanism for this is still up in the air because ligatures must be resolved at the font rendering level. Each font file may provide their own unique ligatures and the mapping from multiple characters to a single glyph can be encoded into a number of different substitution table formats in the font file. Because of this complexity it is often handled by a dedicated library such as [HarfBuzz](https://www.freedesktop.org/wiki/Software/HarfBuzz/). This could of course be wrapped by an R package but since ligatures can not be encoded with unicode (for the most part), there is no way of precomputing the substitution and pass it on to the graphic devices. Whether ligature support can retroactively be added to old graphic devices is thus still up in the air.
```